clients/src/main/java/org/apache/kafka/clients/consumer/StickyAssignor.java

```diff
@@ -278,38 +305,115 @@
         TreeSet<String> sortedCurrentSubscriptions = new TreeSet<>(new SubscriptionComparator(currentAssignment));
         sortedCurrentSubscriptions.addAll(currentAssignment.keySet());
 
-        balance(currentAssignment, sortedPartitions, unassignedPartitions, sortedCurrentSubscriptions,
+        balance(currentAssignment, prevAssignment, sortedPartitions, unassignedPartitions, sortedCurrentSubscriptions,
                 consumer2AllPotentialPartitions, partition2AllPotentialConsumers, currentPartitionConsumer);
         return currentAssignment;
     }
 
+    private void updatePrevAssignment(Map<TopicPartition, ConsumerGenerationPair> prevAssignment,
+                                      TopicPartition partition, ConsumerGenerationPair consumerGeneration) {
+        if (!prevAssignment.containsKey(partition))
+            prevAssignment.put(partition, consumerGeneration);
+        else {
+            ConsumerGenerationPair existingPair = prevAssignment.get(partition);
+            if (consumerGeneration.generation > existingPair.generation)
+                prevAssignment.put(partition, consumerGeneration);
+            else if (consumerGeneration.generation == existingPair.generation) {
+                // same partition is assigned to two consumers during the same rebalance.
+                // log an error and skip this record
+                log.error("Partition '" + partition + "' had been assigned to multiple consumers " +
```

Comment:
nit: Use `{}` substitutions. Maybe this could be `warn` since there's nothing the user can actually do about it shouldn't actually cause any problems for the current assignment.